### PR TITLE
Support for terminal profiles. 

### PR DESCRIPTION
**Pull request by tsmaeder**
#### What it does
Adds support for terminal profiles. 

Fixes [#11503](https://github.com/eclipse-theia/theia/issues/11503)

* UI and service to manage terminal profiles
* Handle profiles in preferences according to VS Code schema
* API and contribution markup for contributing profiles and activation event handling

contributed on behalf of STMicroelectronics

#### How to test
This PR adds two new items under the "Terminal" menu: "New Terminal...", which allows to select a profile and use it to open a terminal and "Choose Default Profile", which selects the profile to be used when selecting "New Terminal" from the menu.
Some profiles are added by default ("cmd" by default on windows), but they can also be contributed by plugins or defined in the preferences. 
Here are two extensions that can be used to test contributed terminals:

[plugins.zip](https://github.com/eclipse-theia/theia/files/10409008/plugins.zip)

One contributes a shell terminal profile ("JavaScript Debug Terminal") and the other a pseudo-pty terminal ("My Extension Terminal"). 

Adding profiles via terminals is described here:

https://code.visualstudio.com/docs/terminal/profiles

Note that terminal "source"s have not been implemented. Theia will accept the syntax, but the resulting profiles will not work.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:92]: These commands do not seem to be present in VS Code, and are not added to the `terminal` main-menu.
I do see the command `'Create New Terminal (With Profile)'` however.

File: packages/terminal/src/browser/terminal-preferences.ts
Comment (vince-fugnitto) [inline packages/terminal/src/browser/terminal-preferences.ts:41]: minor:

```suggestion
        markdownDescription: nls.localize('theia/terminal/profileIcon', 'A codicon ID to associate with the terminal icon. \nterminal-tmux:"$(terminal-tmux)"')
```

**Review by vince-fugnitto — COMMENTED** [review]
It seems that there are conflicts, do you mind rebasing?

File: packages/core/i18n/nls.de.json
Comment (tsmaeder) [reply packages/core/i18n/nls.de.json:467]: Would you prefer translations by a German speaker or machine translations ;-)? As for the original nls.json: I assume the "defaults" would be taken from the code, right? If so, why do we even have the file?

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:92]: terminal-frontend-contribution.ts L 653-668?

File: packages/terminal/src/browser/terminal-preferences.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:41]: What's the difference? Looks identical to me 🤷

File: packages/core/i18n/nls.de.json
Comment (msujew) [reply packages/core/i18n/nls.de.json:467]: Keeping the german translation in there is reasonable, however you should remove the entries from the `nls.json`. The automatic translation update will not be performed in case the `nls.json` doesn't change after performing the extraction CLI command.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (msujew) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:261]: Suggestion: Given that the terminal runs on the backend, it makes more sense to decide on the preference by using the backend OS, instead of the frontend OS.

```suggestion
            switch (OS.backend.type()) {
```

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (msujew) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:280]: Suggestion: Use `backend.isWindows` instead.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (msujew) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:303]: Suggestion:

```suggestion
        switch (OS.backend.type()) {
```

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (msujew) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:971]: Suggestion:

```suggestion
        this.preferenceService.set(`terminal.integrated.defaultProfile.${OS.backend.type().toLowerCase()}`, result[0], PreferenceScope.User);
```

File: packages/terminal/src/browser/terminal-widget-impl.ts
Comment (msujew) [inline packages/terminal/src/browser/terminal-widget-impl.ts:491]:
```suggestion
            shell: this.options.shellPath || this.shellPreferences.shell[OS.backend.type()],
            args: this.options.shellArgs || this.shellPreferences.shellArgs[OS.backend.type()],
```

**Review by msujew — COMMENTED** [review]
I noticed a few issues when the frontend OS isn't the same as the backend OS. The preferences don't match up well.

File: packages/terminal/src/browser/terminal-preferences.ts
Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-preferences.ts:41]: No problem, the extra whitespace.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:92]: I mean, these commands do not exist in VS Code, are they necessary?
Instead we should probably implement `Create New Terminal (With Profile)` which is present in VS Code and likely achieves what we are attempting here.

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:92]: Yes, the UI (in particular choosing the default profile) is very different from what VS Code does. As for the "Create New Terminal (With Profile)", I would argue that my solution is superior: the "..." at the end of "New Profile" is a well established UI pattern indicating that the user will face a dialog to do some configuration or selection before running the action.
Remember that "do the same as VS Code" is not a strategic goal of the project anymore.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:261]: That's not a suggestion: it's a bug.

File: packages/core/i18n/nls.de.json
Comment (tsmaeder) [reply packages/core/i18n/nls.de.json:467]: Thanks for the insight.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:92]: I don't really understand why "New Terminal" should not exist in VS Code: it exists and is under the "terminal" main menu. If you're saying we should use the same command ids, I'm not sure about that: are the command id's supposed to be API in VS Code?

File: packages/terminal/src/browser/terminal-preferences.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:41]: Ah, I see.

File: packages/terminal/src/browser/terminal-preferences.ts
Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:70]: I understand that the default localization cannot handle a computed _key_, but shouldn't it just treat the default value as a random string?

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:70]: I mean: your statement would imply that we cannot default-localize any messages with parameters, right?

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:70]: Ah, I get it.

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-preferences.ts:70]: Ah...you mean the `+`?

Comment (msujew) [reply packages/terminal/src/browser/terminal-preferences.ts]: > Ah...you mean the `+`?

Yes, exactly :)

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: However in this case it adds entries in the `terminal` menu that downstream applications might not be expecting since in general we do follow the default implementation of VS Code. Moreover, plugins may attempt to trigger the `workbench.action.terminal.newWithProfile` command but it does not exist.

If we want to keep the commands that you've added on top of VS Code I'd maybe suggest we move them out of the `terminal` menu.

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: > However in this case it adds entries in the terminal menu that downstream applications might not be expecting

Having the same menu structure as VS Code is not a promise I think we are making to our adopters. If downstream apps rely on that, we have a problem. Our UI with terminals and profiles is very different (no tabbed terminals, for example) and I feel it's a bit of a kludge in VS Code. It would be wrong to strive for visual compatibility in this respect, IMO.

> Moreover, plugins may attempt to trigger the workbench.action.terminal.newWithProfile command but it does not exist.

That's the question I'm asking: is this command id VS Code published API we should support?

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: These are, though: https://code.visualstudio.com/api/references/commands

Comment (vince-fugnitto) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: > That's the question I'm asking: is this command id VS Code published API we should support? I don't think it is.

I believe it is the main reason that command ids follow VS Code, since they can be ultimately executed from plugins.
In the case that we do not match their id, we have a mapping performed in the plugin system, for example:
- https://github.com/eclipse-theia/theia/blob/d383fe25ee6dd2d0a3296222235b73e425666e2f/packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts#L373-L375

I understand you might be referring to [builtin commands](https://code.visualstudio.com/api/references/commands), but the page also describes "common commands" and such commands are also accessible by plugins.

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: Having a look at the broader picture here: all the other terminal-related commands do _not_ have the same command ids as in VS Code (e.g. `terminal:new` instead of `workbench.action.terminal.new`. So having just the two new one's aligned seems odd. In this PR, the command ids are consistent with the rest of the terminal commands. If we want to align the command ids with VS Code, it should be done in a separate PR that covers all terminal-related command ids.

Comment (tsmaeder) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts:91]: I've opened an issue: https://github.com/eclipse-theia/theia/issues/12084

**Review by msujew — APPROVED** [review]
The changes look good to me. I would wait to merge until @vince-fugnitto has given his approval as well 👍

**Review by vince-fugnitto — COMMENTED** [review]
The changes work well for me when I verified the functionality, I only have a minor comment regarding the new menu item.

**Review by vince-fugnitto — APPROVED** [review]
The latest updates look good to me :+1: 
I confirmed that the functionality works well and as expected, and the new API is marked as supported.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts
Comment (colin-grant-work) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts:942]: @tsmaeder, I believe this should have just been `modify({cwd})`. `rootURI` is not a recognized field when these options are eventually passed into the TerminalWidgetFactory.

File: packages/terminal/src/browser/shell-terminal-profile.ts
Comment (colin-grant-work) [inline packages/terminal/src/browser/shell-terminal-profile.ts:39]: This operation is mostly bogus, because `IShellTerminalServerOptions` (the function arg) and `TerminalWidgetOptions` (the constructor argument type) have very little in common.

**Comment by tsmaeder** [comment]
I think this is a leftover from prototype I did where I separated the code paths for shell terminals and other terminals (`Pseudoterminal`).

**Pull request merged**
